hail/python/hail/linalg/utils/misc.py

```diff
@@ -165,67 +167,58 @@ def locus_windows(locus_expr, radius, coord_expr=None):
-        global_pos = np.zeros(n_loci, dtype=np.int64)
-        for i, p in enumerate(global_pos_list):
-            if p is None:
-                raise ValueError(f"locus_windows: missing value for 'locus_expr' global position at row {i}")
-            global_pos[i] = p
-        coord = global_pos
-        del global_pos_list
-    else:
+    if coord_expr is not None:
         check_row_indexed('locus_windows', coord_expr)
-        global_pos_and_coord =\
-            hl.tuple([locus_expr.global_position(), coord_expr]).collect()  # raises exception if sources differ
-        n_loci = len(global_pos_and_coord)
-
-        global_pos = np.zeros(n_loci, dtype=np.int64)
-        coord = np.zeros(n_loci, dtype=np.float64)
-        for i, x in enumerate(global_pos_and_coord):
-            if x[0] is None:
-                raise ValueError(f"locus_windows: missing value for 'locus_expr' global position at row {i}")
-            global_pos[i] = x[0]
-            if x[1] is None:
-                raise ValueError(f"locus_windows: missing value for 'coord_expr' at row {i}")
-            coord[i] = x[1]
-        del global_pos_and_coord
-
-    if n_loci == 0:
-        return np.zeros(shape=0, dtype=np.int64), np.zeros(shape=0, dtype=np.int64)
 
-    contig_name = locus_expr.dtype.reference_genome.contigs
-    contig_len = locus_expr.dtype.reference_genome.lengths
-    contig_cum_len = np.cumsum([contig_len[name] for name in contig_name])
+    src = locus_expr._indices.source
+    if locus_expr not in src._fields_inverse:
+        locus = Env.get_uid()
+        annotate_fields = {locus: locus_expr}
 
-    assert(global_pos[-1] < contig_cum_len[-1])
+        if coord_expr is not None:
+            if coord_expr not in src._fields_inverse:
+                coords = Env.get_uid()
+                annotate_fields[coords] = coord_expr
+            else:
+                coords = src._fields_inverse[coord_expr]
 
-    contig_start_idx = _compute_contig_start_idx(global_pos, contig_cum_len)
-    n_contigs = len(contig_start_idx)
-    contig_start_idx.append(n_loci)
-    contig_bounds = [array_windows(coord[contig_start_idx[c]:contig_start_idx[c + 1]], radius)
-                     for c in range(n_contigs)]
-    starts = np.concatenate([contig_start_idx[c] + contig_bounds[c][0] for c in range(n_contigs)])
-    stops = np.concatenate([contig_start_idx[c] + contig_bounds[c][1] for c in range(n_contigs)])
+        if isinstance(src, hl.MatrixTable):
+            new_src = src.annotate_rows(**annotate_fields)
+        else:
+            new_src = src.annotate(**annotate_fields)
 
-    return starts, stops
+        locus_expr = new_src[locus]
+        if coord_expr is not None:
+            coord_expr = new_src[coords]
+
+    if coord_expr is None:
+        coord_expr = locus_expr.position
+
+    rg = locus_expr.dtype.reference_genome
+    contig_group_expr = hl.agg.group_by(hl.locus(locus_expr.contig, 1, reference_genome=rg), hl.agg.collect(coord_expr))
```

Comment:
sorry to keep going back and forth on this PR, but this stuff is a bit complicated. 

This is a locus instead of a contig because we want it to sort in contig order below, right? At some point we can add a `hl.contig_index(contig, rg)` function which would let us just take the contig, but don't need to do that here.

Reply:
yeah, for sure! That would be a good addition. (I considered doing a bit of a hack to do essentially that here, but decided against it. Adding a function on the reference genome would be a good way to do it.)

Also you meant to approve, yeah? Because it just merged.

Reply:
yeah, this was a non-blocking discussion point.